**Tarea 1: Análisis financiero de operaciones diarias usando Python**

**Objetivo:** Reforzar tu comprensión sobre las estructuras de datos y control en Python mediante el análisis de las operaciones diarias de una empresa.

**Descripción** Eres un analista financieroen una empresa de distribución de productos electrónicos. Cada día, la empresa tiene múltiples transacciones que involucran la compra de productos a proveedores y la venta a clientes. Se te ha asignado la tarea de analizar las operaciones del último mes para obtener insights sobre las operaciones de la empresa.



### Datos a usar
El siguiente código genera datos de 600 operaciones de compra y venta realizadas en agosto. Los guarda en `generated_operaciones`  que es una lista de diccionarios (una lista donde cada entrada es un diccionario).

Cada operación tiene los datos fecha de operación, tipo de operación (compra o venta), codigo de producto, cantidad y precio_unitario.




In [1]:
import random
from datetime import datetime, timedelta
random.seed(42)

num_observations = 600
generated_operaciones = []

for _ in range(num_observations):
    fecha = (datetime.strptime("2023-08-01", "%Y-%m-%d") + timedelta(days=random.randint(0, 30))).strftime('%Y-%m-%d')
    tipo = random.choice(["compra", "venta"])
    codigo_de_producto = "P00" + str(random.randint(1, 3))
    cantidad = random.randint(1, 20)
    precio_unitario = random.randint(100, 1000)

    operacion = {
        "fecha": fecha,
        "tipo": tipo,
        "codigo_de_producto": codigo_de_producto,
        "cantidad": cantidad,
        "precio_unitario": precio_unitario
    }

    generated_operaciones.append(operacion)


In [2]:
generated_operaciones[:5]

[{'fecha': '2023-08-21',
  'tipo': 'compra',
  'codigo_de_producto': 'P001',
  'cantidad': 9,
  'precio_unitario': 350},
 {'fecha': '2023-08-08',
  'tipo': 'compra',
  'codigo_de_producto': 'P003',
  'cantidad': 4,
  'precio_unitario': 792},
 {'fecha': '2023-08-24',
  'tipo': 'compra',
  'codigo_de_producto': 'P003',
  'cantidad': 14,
  'precio_unitario': 132},
 {'fecha': '2023-08-01',
  'tipo': 'compra',
  'codigo_de_producto': 'P001',
  'cantidad': 8,
  'precio_unitario': 617},
 {'fecha': '2023-08-20',
  'tipo': 'compra',
  'codigo_de_producto': 'P003',
  'cantidad': 7,
  'precio_unitario': 833}]

### **Continúa desde aquí**
#### **Tarea:**
La primera está casi completa (falta completar), úsala como guía por si estás perdido en cómo empezar. Puedes modificar el código como quieras.

**Agrega comentarios o texto si crees necesario para explicar tu código**

1. Calcula el total de operaciones de compra y venta realizadas durante el mes **para el producto P003**.




In [3]:
# Inicializamos los contadores
total_compras = 0
total_ventas = 0

# Iteramos sobre las transacciones
for transaccion in generated_operaciones:
    if transaccion["tipo"] == "compra" and transaccion["codigo_de_producto"] == "P003":
        total_compras += 1
    elif transaccion["tipo"] == "venta" and transaccion["codigo_de_producto"] == "P003":
        total_ventas += 1

total_compras

97

In [4]:
total_ventas

111


2. Determina la cantidad total de productos (por código) comprados y vendidos durante el mes.  

In [5]:
def producto(codigo, lista):
    venta = 0
    compra = 0
    
    for x in lista:
        if x['tipo'] == 'compra' and x['codigo_de_producto'] == codigo:
            compra = compra + x['cantidad']
        elif x['tipo'] == 'venta' and x['codigo_de_producto'] == codigo:
            venta = venta + x['cantidad']
    
    return (venta,compra)

In [6]:
p001 = producto("P001", generated_operaciones)
p002 = producto("P002", generated_operaciones)
p003 = producto("P003", generated_operaciones)

print(p001,p002,p003)

(1202, 915) (1105, 926) (1265, 1037)



3. Identifica el producto más vendido durante el mes.

El producto más vendido fue el `P003`

4. Calcula el precio de venta promedio del producto P001.

In [7]:
denominador = 0
contador = 0

for i in generated_operaciones:
    if i['codigo_de_producto']=="P001" and i['tipo']=="venta":
        denominador = denominador + i['cantidad']*i['precio_unitario']
        contador = contador + i['cantidad']

precio_vta_promedio = round(denominador/contador,2)
print(precio_vta_promedio)

558.82



5. Calcula las ganancias por producto.

In [8]:
def ganancias(codigo, lista):
    venta = 0
    compra = 0
    
    for x in lista:
        if x['tipo'] == 'compra' and x['codigo_de_producto'] == codigo:
            compra = compra + x['cantidad']*x['precio_unitario']
        elif x['tipo'] == 'venta' and x['codigo_de_producto'] == codigo:
            venta = venta + x['cantidad']*x['precio_unitario']
    
    return venta-compra

In [9]:
p001 = ganancias("P001", generated_operaciones)
p002 = ganancias("P002", generated_operaciones)
p003 = ganancias("P003", generated_operaciones)

print(p001,p002,p003)

150716 65765 133982


6. Libre: Desarrolla un insight libre, propuesto por ti

Cálculo del ticket promedio

In [10]:
acumulado = 0
for i in generated_operaciones:
    acumulado = acumulado + i['precio_unitario']*i['cantidad']

acumulado/len(generated_operaciones)

6045.721666666666

El ticket promedio de las operaciones es de **$6,045.72$**